### ISSUE TITLE
Epic: Linux Containers on Windows (LCOW)

### ISSUE THREAD

**Issue by PatrickLang** [moby/moby#33850]
Microsoft is hard at work adding support to run both Windows & Linux containers side by side on the same node with a single Docker daemon. As Linux containers are launched, Hyper-V will be used to boot and run a Linux kernel which will then be used to host the container natively. Because many areas of Docker will require changes to handle running multiple platforms side by side, we'll be using this epic to track the list of PRs and proposals in a single place.

## Preview documentation
https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/linux-containers 

## Progress

**Required for Docker for Windows - Developer Scenarios**

Status        | Description
--------------|------------------------------------------------------------------
[Merged](https://github.com/moby/moby/pull/33241)   | Store Linux images on Windows & launch Linux containers on Window. Right now, platform switching is done with a system-wide environment variable. Side by side support needs the next item.
[Merged](https://github.com/moby/moby/pull/33969) | Support Linux filesystem operations in single-use VMs instead of reusable ones. This can provide protection in the case where a file parsing bug could lead to a container polluting the filesystem of another container.
[Merged](https://github.com/moby/moby/pull/34568) | Use a single tag store so both Windows & Linux containers can be stored and managed by the same engine
[Merged](https://github.com/moby/moby/pull/34252)  | Remote filesystem support needed for `docker commit`, `docker cp`, and `Dockerfile` `ADD` commands
[Merged](https://github.com/moby/moby/pull/34170) |  Support resizing ext4 sandbox VHD instead of defaulting to 20Gb only. 
[Merged](https://github.com/moby/moby/pull/34258)  | Bind mount support.
[Merged](https://github.com/moby/moby/pull/34642)   | New `--platform` parameter for `docker run`, `docker pull`, `docker import` and `docker build` to choose between platforms and order of precedence when multiple are supported (Windows & Linux) on the same node. See Proposal #34617 .
[Merged](https://github.com/moby/moby/pull/34859) | Coalesce daemon stores - needed to manage both Windows & Linux containers with the same daemon
In progress | Add `docker kill -s` for signals != SIGKILL. 
[Merged](https://github.com/moby/moby/pull/34405) | Support `WORKDIR` in running existing images and building new ones
Not started   | Add platform filters to `docker search` and `docker images`
[Merged](https://github.com/moby/moby/pull/37296)| Memory & CPU settings - see: https://github.com/Microsoft/opengcs/issues/145
Not started | LCOW: `docker run --user` seems to be ignored https://github.com/moby/moby/issues/36469

**Required for Kubernetes**

Status        | Description
--------------|------------------------------------------------------------------
Not started   | Update `docker stats` for multiple platforms
Proposal soon | Update `docker info` & daemon configuration to show what platforms each node is capable of, and restrict to single platforms if needed

**Required for Swarm**

Status        | Description
--------------|------------------------------------------------------------------
Not started   | Add `--platform` flag support to `docker-compose`, and `docker stack deploy`
Not started   | Update `docker stats` for multiple platforms
Not started   | Support multiple platforms for `docker volume` commands
Not started   | Adjust Swarm mode placement to handle running multiple platforms instead of just the node's native platform. Until this is done, use `docker service create --no-resolve-image` as a workaround

**Backlog, need to set priority**

Status        | Description
--------------|------------------------------------------------------------------
Not started | `docker top`
Not started | `docker run` flags. Feedback welcome on what's needed vs not
Not started | `docker export`, `docker import`, `docker save` need platform flags
Not started | UID & GID flag support - see: https://github.com/Microsoft/opengcs/issues/146
No plans to support | Cross-platform multi stage builds, support for `COPY` command in `Dockerfile`

## Glossary & References

- LCOW - **L**inux **C**ontainers **o**n **W**indows
- GCS - guest compute service. Receives commands from the Hyper-V host services (DockerD via HCS) and invokes processes for specific tasks, creating namespaces, and creating containers using runC. (Repo link coming soon)
- HCS - host compute service. Windows-specific service used to manage containers and VMs (doc link coming soon)

There's a brief description of gcs & hcs from a [Dockercon 2016 presentation](https://youtu.be/85nCF5S8Qok?t=421)

## FAQs

What's needed to fully test this?

- As of https://github.com/moby/moby/pull/39108, Windows 10 or Windows Server version 1809 (build 17763), or a later Windows Insider build
- A build of moby/moby with the PR's above applied
- A bootable UEFI-based Linux image with the right modules & gcs included - see https://github.com/microsoft/opengcs 
- At least 4GB of RAM and Intel VT or AMD-V instruction support
  - Note: other apps using these instructions (VMWare, Virtualbox, ...) cannot be run concurrently

**Comment by AkihiroSuda** [comment]
What would be a usecase of "Cross-platform multi stage builds"?

**Comment by PatrickLang** [comment]
@AkihiroSuda It would be great for cross platform builds. For example, if you had a collection of bash scripts and a language with cross-compilation support like golang, then you could actually build your code in Linux then copy the binaries into a `microsoft/nanoserver` image to run it in a Windows native container. The same would be feasible for projects based on JVM, .Net Core, or Mono.

**Comment by simonferquel** [comment]
Just updated my insider machine to 16257, and the bind mount PR still fails:
`docker run -it --rm -v c:\Users:/data busybox`
```
docker: Error response from daemon: container 0c01e8361b77d03bfe048812b7e012e23334201aa0660e698687db1bbf9f6d08 encountered an error during CreateProcess: failure in a Windows system call: Unspecified error (0x80004005) extra info: {"CommandArgs":["sh"],"WorkingDirectory":"/","Environment":{"HOSTNAME":"0c01e8361b77","PATH":"/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin","TERM":"xterm"},"EmulateConsole":true,"CreateStdInPipe":true,"CreateStdOutPipe":true,"ConsoleSize":[30,120],"OCISpecification":{"ociVersion":"1.0.0-rc5-dev","platform":{"os":"linux","arch":"amd64"},"process":{"terminal":true,"consoleSize":{"height":30,"width":120},"user":{"uid":0,"gid":0},"args":["sh"],"env":["PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin","HOSTNAME=0c01e8361b77","TERM=xterm"],"cwd":"/","capabilities":{"bounding":["CAP_CHOWN","CAP_DAC_OVERRIDE","CAP_FSETID","CAP_FOWNER","CAP_MKNOD","CAP_NET_RAW","CAP_SETGID","CAP_SETUID","CAP_SETFCAP","CAP_SETPCAP","CAP_NET_BIND_SERVICE","CAP_SYS_CHROOT","CAP_KILL","CAP_AUDIT_WRITE"],"effective":["CAP_CHOWN","CAP_DAC_OVERRIDE","CAP_FSETID","CAP_FOWNER","CAP_MKNOD","CAP_NET_RAW","CAP_SETGID","CAP_SETUID","CAP_SETFCAP","CAP_SETPCAP","CAP_NET_BIND_SERVICE","CAP_SYS_CHROOT","CAP_KILL","CAP_AUDIT_WRITE"],"inheritable":["CAP_CHOWN","CAP_DAC_OVERRIDE","CAP_FSETID","CAP_FOWNER","CAP_MKNOD","CAP_NET_RAW","CAP_SETGID","CAP_SETUID","CAP_SETFCAP","CAP_SETPCAP","CAP_NET_BIND_SERVICE","CAP_SYS_CHROOT","CAP_KILL","CAP_AUDIT_WRITE"],"permitted":["CAP_CHOWN","CAP_DAC_OVERRIDE","CAP_FSETID","CAP_FOWNER","CAP_MKNOD","CAP_NET_RAW","CAP_SETGID","CAP_SETUID","CAP_SETFCAP","CAP_SETPCAP","CAP_NET_BIND_SERVICE","CAP_SYS_CHROOT","CAP_KILL","CAP_AUDIT_WRITE"]}},"root":{"path":"rootfs"},"hostname":"0c01e8361b77","mounts":[{"destination":"/proc","type":"proc","source":"proc","options":["nosuid","noexec","nodev"]},{"destination":"/dev","type":"tmpfs","source":"tmpfs","options":["nosuid","strictatime","mode=755"]},{"destination":"/dev/pts","type":"devpts","source":"devpts","options":["nosuid","noexec","newinstance","ptmxmode=0666","mode=0620","gid=5"]},{"destination":"/sys","type":"sysfs","source":"sysfs","options":["nosuid","noexec","nodev","ro"]},{"destination":"/sys/fs/cgroup","type":"cgroup","source":"cgroup","options":["ro","nosuid","noexec","nodev"]},{"destination":"/dev/mqueue","type":"mqueue","source":"mqueue","options":["nosuid","noexec","nodev"]},{"destination":"/data","source":"c:\\users"}],"linux":{"resources":{"devices":[{"allow":false,"access":"rwm"},{"allow":true,"type":"c","major":1,"minor":5,"access":"rwm"},{"allow":true,"type":"c","major":1,"minor":3,"access":"rwm"},{"allow":true,"type":"c","major":1,"minor":9,"access":"rwm"},{"allow":true,"type":"c","major":1,"minor":8,"access":"rwm"},{"allow":true,"type":"c","major":5,"minor":0,"access":"rwm"},{"allow":true,"type":"c","major":5,"minor":1,"access":"rwm"},{"allow":false,"type":"c","major":10,"minor":229,"access":"rwm"}]},"namespaces":[{"type":"mount"},{"type":"network"},{"type":"uts"},{"type":"pid"},{"type":"ipc"}],"maskedPaths":["/proc/kcore","/proc/latency_stats","/proc/timer_list","/proc/timer_stats","/proc/sched_debug"],"readonlyPaths":["/proc/asound","/proc/bus","/proc/fs","/proc/irq","/proc/sys","/proc/sysrq-trigger"]}}}
```

I suspect I have something special to pass to HCS, so that it creates 9p listeners / mount points

**Comment by AaronFriel** [comment]
Not sure where to comment on this (so many PRs!), but the Docker and Ubuntu tutorials are not working with the 2017-09-16 binaries from https://master.dockerproject.org/.

This is the error:

```
PS ...> .\docker.exe run -ti busybox sh
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
C:\Program Files\Docker\Docker\Resources\bin\docker.exe: no matching manifest for windows/amd64 in the manifest list entries.
See 'C:\Program Files\Docker\Docker\Resources\bin\docker.exe run --help'.
```

`LCOW_SUPPORTED` is 1:

```
PS ...> echo $ENV:LCOW_SUPPORTED
1
```

Daemon verson:

```
> .\dockerd -v
Docker version master-dockerproject-2017-09-16, build c5c0702
```

Client version:

```
> .\docker -v
Docker version master-dockerproject-2017-09-16, build 6d614ec
```

**Comment by friism** [comment]
@AaronFriel when starting the daemon, did you add the `--experimental flag`?

**Comment by friism** [comment]
If that doesn't work, you can also run a specific sha:

    docker -H "npipe:////./pipe//docker_lcow" run -ti busybox@sha256:b82b5740006c1ab823596d2c07f081084ecdb32fd258072707b99f52a3cb8692 sh

**Comment by AaronFriel** [comment]
@friism It works with `dockerd.exe` and `docker.exe` binaries downloaded today, I see: `Docker version 17.09.0-ce-rc1, build ae21824` on docker.exe, and `a2ee40b` on `dockerd.exe`.

**Comment by friism** [comment]
@AaronFriel great, thanks for following up

**Comment by carlfischer1** [comment]
@PatrickLang please add an entry in the table for supporting the --platform flag in the compose file format, docker-compose, and docker stack deploy

**Comment by galvesribeiro** [comment]
Just a report on LCOW:

If I run:

```
docker run --rm -it -v C:\Users\gutem\source\repos\DGC\DGC.Router.Providers:/app --workdir /app  microsoft/dotnet:sdk bash
```

And then inside the container:

```
root@e6642bee4a71:/app# dotnet restore teste.sln
  Restoring packages for /app/src/DGC.Router.Providers/DGC.Router.Providers.csproj...
  Generating MSBuild file /app/src/DGC.Router.Providers/obj/DGC.Router.Providers.csproj.nuget.g.props.
/usr/share/dotnet/sdk/2.0.3/NuGet.targets(102,5): error : Access to the path is denied. [/app/teste.sln]
/usr/share/dotnet/sdk/2.0.3/NuGet.targets(102,5): error :   Permission denied [/app/teste.sln]
```

I have no idea why we are having permission denied inside the container. If I run it outside the container, it works perfectly.

**Comment by thaoula** [comment]
Can bridge mode networking be added so that we can use the same compose file on windows, mac and linux.

All our compose files reference bridge but this fails when using LCOW because there is no bridge network.

Can an alias be created bridge = nat?

**Comment by bgervin** [comment]
`docker run --rm -v /test/myfile.txt:/myfile.txt bgervin/tabula-java`
Running the above command, I would have expected a myfile.txt file to be in the root at /myfile.txt - however from within the container - myfile.txt is showing up as a directory.

**Comment by AaronFriel** [comment]
Windows Server 2019 preview notes say: [A feedback we constantly hear from developers is the complexity in navigating environments with Linux and Windows deployments. To address that, we previously extended Windows Subsystem on Linux (WSL) into insider builds for Windows Server, so that customers can run Linux containers side-by-side with Windows containers on a Windows Server. In Windows Server 2019, we are continuing on this journey to improve WSL, helping Linux users bring their scripts to Windows while using industry standards like OpenSSH, Curl & Tar.](https://cloudblogs.microsoft.com/windowsserver/2018/03/20/introducing-windows-server-2019-now-available-in-preview/)

Does this mean WSL will be capable of running containers? If so, is there any guidance at this point?

**Comment by edrevo** [comment]
@PatrickLang , would it be possible to add to this epic a task to support an easy way of binding the docker pipe / socket to a container? See https://github.com/moby/moby/issues/36574

**Comment by lwsrbrts** [comment]
Spent an evening battling with the shortcomings of LCOW on Windows Server 1803 attempting to start an nginx reverse proxy that uses "secrets" and "configs" for configuration and there's basically no solution other than to embed the data by creating a Dockerfile...or just creating a Linux VM!

Attempts:

1. Create a YAML and `docker stack deploy`.
2. Try using version: '3.7' and '2.4' on compose file - `3.7` doesn't support the `platform` setting, `2.4` can't use configs or secrets.
4. `docker-compose up` has the same issues.
5. `docker container run` doesn't have a `platform` switch
6. Pre-downloading the image doesn't work.
7. `docker service create` doesn't have a `platform` switch

Appreciate it's a work in progress but I'm eager to see this in action.

**Comment by lwsrbrts** [comment]
@olljanat - I'm talking about Windows Server 1803 and Docker EE preview with the experimental flag, so my configuration does contain these features and in fact I'm sure that's what this article is all about or presumably this epic would be in the down-level docker-for-windows repository.

Finding documentation that clearly states what is and isn't supported when the feature is experimental and very much a work in progress is troublesome of course and only after I was getting issues did it become necessary to look and is when I found this article.

To be as clear as possible, I'm trying things out and seeing where the limits are and then reporting my findings - presumably part of the whole development process.

I've configured Windows Server 1803 to allow me to run this: `docker container run -d --platform linux -p 80:80 nginx` and the container will start and work.

This is roughly what I used to enable LCOW on Windows Server 1803.

```powershell
Install-WindowsFeature -Name Containers -IncludeManagementTools

Restart-Computer

Install-Module DockerMsftProvider -Force

Install-Package Docker -ProviderName DockerMsftProvider -Force

Set-Location $env:ProgramFiles\Docker

Start-BitsTransfer -Destination dockerd.exe -Source https://master.dockerproject.org/windows/x86_64/dockerd.exe
Start-BitsTransfer -Destination docker.exe -Source https://master.dockerproject.org/windows/x86_64/docker.exe


[Net.ServicePointManager]::SecurityProtocol = "tls12, tls11, tls"
$wc = New-Object System.Net.WebClient
$wc.DownloadFile('https://github.com/linuxkit/lcow/releases/download/4.14.29-0aea33bc/release.zip', "C:\release.zip")

Invoke-WebRequest "https://github.com/docker/compose/releases/download/1.22.0/docker-compose-Windows-x86_64.exe" -UseBasicParsing -OutFile $Env:ProgramFiles\docker\docker-compose.exe

sc.exe delete Docker

Expand-Archive c:\release.zip -DestinationPath "$Env:ProgramFiles\Linux Containers\"

Copy-Item "$Env:ProgramFiles\Linux Containers\bootx64.efi" "$Env:ProgramFiles\Linux Containers\kernel"

. $env:ProgramFiles\docker\dockerd.exe --register-service --experimental

[psobject]@{"data-root" = 'F:\Containers' } | ConvertTo-Json | Out-File -Encoding default -FilePath $env:ProgramData\docker\config\daemon.json

Start-Service docker

docker run --platform linux --rm -ti busybox sh

docker run --platform linux --rm -d --name nginx -p 80:80 nginx
```

Which gives me this output from a `docker info`

```
PS F:\> docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 4
Server Version: master-dockerproject-2018-09-08
Storage Driver: windowsfilter (windows) lcow (linux)
 Windows:
 LCOW:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd gcplogs gelf json-file local logentries splunk syslog
Swarm: active
 NodeID: t5x5yhgh0ef41b4w0ne5m3c2h
 Is Manager: true
 ClusterID: jnhpdawqs60exns1zoseo6kgp
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 192.168.1.4
 Manager Addresses:
  192.168.1.4:2377
Default Isolation: process
Kernel Version: 10.0 17134 (17134.1.amd64fre.rs4_release.180410-1804)
Operating System: Windows Server Datacenter Version 1803 (OS Build 17134.228)
OSType: windows
Architecture: x86_64
CPUs: 40
Total Memory: 63.69GiB
Name: HYPERV2
ID: RIDJ:WAGS:4ELS:FPS6:EZXK:UD7P:CVSF:YH7E:KGFF:7ER4:YTWH:HQY7
Docker Root Dir: F:\Containers
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

PS F:\>
```

And this is the contents of the "Linux Containers" folder.

```
PS C:\Program Files\Linux Containers> dir


    Directory: C:\Program Files\Linux Containers


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       23/03/2018     18:22        7656208 bootx64.efi
-a----       23/03/2018     18:22        6449925 initrd.img
-a----       23/03/2018     18:22        7656208 kernel
-a----       23/03/2018     18:22            113 versions.txt


PS C:\Program Files\Linux Containers> cat .\versions.txt
linuxkit version v0.2+
commit: 2dd0c9f63d2014000a22ba4e684e3a7f6105eba1
a368a53c78f1491afbbfd3f483ed84d83a5295cb
```

**Comment by vishalniit** [comment]
Since Windows Version 1809 (OS Build 17763.134) is out, does anybody know if --platform flag works with docker-compose file ?

**Comment by kendallroden** [comment]
What is the state of this? Is this still a preview-only feature for windows server 2016? I just spun up a VM and would like to experiment with LCOW

**Comment by thaJeztah** [comment]
Linux Containers On Windows ("LCOW") is still experimental, and not yet feature complete; but you can run it / experiment with it.

**Comment by kendallroden** [comment]
@thaJeztah I have no prob doing that on Docker for Windows in Windows 10, but I know it is not a recommended practice to install docker for windows on Windows server.  Is there a way to run the LCOW on docker ee that is natively baked into Windows Server 2016?

**Comment by olljanat** [comment]
@kendallroden I'm not LCOW user but afaik enterprise and community editions are very near of each others on 18.09. How ever I would highly recommend to use Windows Server 2019 because 2016 have a lot of known limitations/issues which will not be ever fixed.

Btw. I would appreciate if you can share on MicrosoftDocs/Virtualization-Documentation#851 which kind of use cases you would use this if/when it is ready?

**Comment by thaJeztah** [comment]
> afaik enterprise and community editions are very near of each others on 18.09

Docker Desktop ("Docker for Windows") in general is targeted at "developers", and not "production" purposes; Docker Desktop contains a Docker Community engine and cli, but in addition to that, bundles;

- A "moby" VM with LinuxKit, in which a Docker Engine is running to run Linux Containers
- A Windows Docker Daemon to run native Windows containers (or to use LCOW)
- A "one-click" option to install a local Kubernetes for local development
- Other tools, such as `docker-compose`, `docker machine`
- A tray-icon / control-panel to make configuration changes to the docker engine, the "moby" VM, and to switch between "Linux" and "Windows" containers

Docker Desktop allows you to either run Linux containers inside the "moby" VM (giving all the options that a regular Docker Engine running on Linux has), or use LCOW to run Linux containers on the Windows daemon.

Docker Engine Enterprise is targeted at production use, and does not have all the "bells and whistles" of Docker Desktop that are targeted at developers (no "moby" VM, no docker-compose or docker-machine bundled, and no "one-click Kubernetes"). Docker Engine Enterprise comes with Long term support, a number of EE-only features (FIPS support, multiple logging-driver support, Daemon-side Docker Trust), as well options for commercial support.

For Linux, Docker Engine is a paid product; for Windows Server, is available at no additional cost; a basic license comes free with Windows Server (with primary support provided by Microsoft)

Docker Engine Enterprise should also be able to run LCOW (if experimental features are enabled on the daemon), but no support is provided for experimental features ("things may break: here be dragons").

(also see https://www.docker.com/products/docker-engine and https://hub.docker.com/editions/enterprise/docker-ee-server-windows)

**Comment by olljanat** [comment]
Thanks for clarifying. What I was meaning the fact that earlier there was own release notes page for Engine Enterprise and Community Edition but now they both point to [this](https://docs.docker.com/engine/release-notes/) common one which says:
> NOTE: New in 18.09 is an aligned release model for Docker Engine - Community and Docker Engine - Enterprise. The new versioning scheme is YY.MM.x where x is an incrementing patch version. The enterprise engine is a superset of the community engine. They will ship concurrently with the same x patch version based on the same code base.

Reason why I have been asking of this and on different issues about where people are planning to use LCOW is that IMO it puts Microsoft focus to totally wrong place. Like you say Windows Server comes with Docker Engine Enterprise (which should be planned for production use):
> For Linux, Docker Engine is a paid product; for Windows Server, is available at no additional cost; a basic license comes free with Windows Server (with primary support provided by Microsoft)

And on reality there is lot of new bugs on Win Srv 2019 which prevents customers to upgrade their production hosts to it:
https://github.com/moby/moby/issues/38114#issuecomment-437956101
#38358
#38484

Windows version of Moby also misses Swarm integrations tests which is most probably the reason why we have #38484 open now and those are very boring for customers like us who really have Windows containers on production use.

And on other side there is solutions like [RancherOS](https://rancher.com/rancher-os/) which provides fully working, production ready Docker engine with minimal overhead (all our production Linux Docker nodes are using it). So IMO LCOW support for Windows Server should be put to category "will not be done".

PS. Sorry, this post is bit offtopic but I can see that many people from Microsoft are following this issue so hopefully also some who is able to improve situation will get this feedback.

**Comment by jhowardmsft** [comment]
> Is there a way to run the LCOW on docker ee that is natively baked into Windows Server 2016?

@kendallroden Yes. Although Windows Server 2019 would be highly preferred. If 2016, it must be at least RS3.  You will need a kernel and an initrd.img (eg https://github.com/linuxkit/lcow/releases) installed in `C:\Program Files\Linux Containers`. You must start the daemon in experimental mode `dockerd --experimental`. Then you should be able to do `docker pull --platform=linux busybox`.

**Comment by kendallroden** [comment]
Thank you @jhowardmsft , this is exactly what I was looking for

**Comment by z2z** [comment]
thank you @lwsrbrts , I was able to run [Gogs](https://github.com/gogs/gogs/tree/master/docker) on 1809 (17763.107)

`docker run --name=gogs -p 10022:22 -p 10080:3000 -v d:/gogsdata/var/gogs:/data gogs/gogs
`

**Comment by ggirard07** [comment]
Issue tracking hasn't been updated in year.
Should we consider this experimental feature still under active development or did you move to a new approach in the meantime?

**Comment by jhowardmsft** [comment]
Agreed this hasn't been updated much. However
 - Yes, LCOW is still being developed
 - We are focusing on moving the runtime to containerd (as per a LCOL) daemon, and have invested heavily in LCOW through the HCS "v2" APIs which are used in that runtime. Docker will pick up many of those benefits "for free" once that migration is complete (no, I'm not going to provide any date).
 - Yes, it remains experimental.

**Comment by ggirard07** [comment]
@jhowardmsft thanks for taking time to provide feedback, this is really appreciated.
I totally understand now why dev seemed "stuck" whereas there is a lot currently happening under the hood to ease LCOW support in moby, which seems hard to maintain/evolve right now.

Also totally understand the `no date to provide`. Only feedback I can give you (and perhaps this should be in `Microsoft/hcsshim` and not here) is having a milestone or label associated to HCS "v2" in `Microsoft/hcsshim` repo to track related issues will allow us to have that kind of feedback without poking you every time :)

**Comment by maxwellb** [comment]
_**Docker for Windows - Developer Scenarios:** Memory & CPU settings_ appears to be addressed by moby/moby#37296. Very exciting!

**Comment by jhowardmsft** [comment]
Just want to give a heads-up advance notice re LCOW. There's not a particularly good place to advertise this, but this issue seems as reasonable a place as any. At some point in the future, LCOW (the experimental feature, not mobyVM in Docker-For-Windows) will require RS5 or later (ie build 17763+). RS5 on Server is "Windows Server 2019". Put another way, we will be dropping support on RS3 and RS4 builds. It will still be a while, but note this will be coming. 

Some background as to why.... RS1/Windows Server 2016; RS2 (client only); and RS3 (server and client) only support HCS "v1" APIs in the platform. LCOW as currently written is against those v1 APIs. RS4 introduced (but not completed everything needed) HCS "v2" APIs in the platform. RS5 is the first OS release to have enough support in the platform through the "v2" APIs to support LCOW.

We have been extremely busy in the https://github.com/Microsoft/hcsshim repo so that it is possible for "a runtime" (see below) to use "v2" APIs through an OCI-spec-driven interface. That work extends for both WCOW (in process and Hyper-V isolation varients) as well as LCOW. We have deliberately chosen to not provide support in hcsshim for LCOW using HCS v1 APIs - the support burden is simply too great,  especially when the support-lifetime for Windows 10 RS3 and RS4 expires this year - https://support.microsoft.com/en-us/help/13853/windows-lifecycle-fact-sheet.  (For those wondering, WCOW will still work using HCS v1 for RS1 through hcsshim, as RS1 is an LTSC on server. WCOW on RS5+ will move to use "v2" APIs only).

When I say "a runtime" above, I particularly mean containerd, using the shim from the hcsshim repo as the underlying runtime/thing which talks to Windows itself. Obviously there are several avenues for using containerd - docker is just one of them, and I have recently added experimental/incomplete code to the moby repo to allow Windows daemons to use containerd as a runtime - https://github.com/moby/moby/pull/38541. Other "orchestrators" beside docker include CRI/containerd; kube; .... 

Anyway, too much information. Just wanted to make everyone aware that the LCOW experimental feature will require RS5 at some point later this year.

**Comment by jhowardmsft** [comment]
Thanks @maxwellb - Updated description.

**Comment by ndarlington** [comment]
@jhowardmsft Sorry I'm not too au fait with this release numbering; will this RS5-only move/update also include a Windows 10 RS5+ (for LCOW within Docker Desktop)?  Or Is that going to become end of life, with only Windows Server 2019 (RS5+) supporting or being able to support it in future?

**Comment by kendallroden** [comment]
Can someone confirm that running Linux containers on Windows Server 2019 will still be an "experimental" feature vs. baked in?

**Comment by thaJeztah** [comment]
Yes, still experimental

**Comment by tillig** [comment]
> Cross-platform multi stage builds, support for `COPY` command in `Dockerfile`

Does this mean _same platform multi stage builds_ will also not get `COPY` support? Or is that just for _cross-platform_? Reason I ask, on engine 19.03.0-rc2 in Windows mode with experimental features on... using a `COPY` in a `Dockerfile` to copy from one Linux container to another fails with "file not found" errors. Launching the intermediate containers, you can see the source and target files/folders do all exist, so it seems like I might be hitting this limitation. (Or a bug, depending on what this means.)

**Comment by MZap** [comment]
Any status on the platform flag for Compose?

**Comment by jhowardmsft** [comment]
@mzap probably a question to ask over at https://github.com/docker/compose

**Comment by oldium** [comment]
Is the mentioned v2 API related to WSL 2 announced recently for Windows Insiders and to Docker preview, which shall be available in July?

**Comment by jhowardmsft** [comment]
WSL2 uses some of the HCS v2 APIs. But no, there's no relation as such between LCOW in moby and WSL2.

**Comment by oldium** [comment]
So Linux containers support in Docker preview (which is not yet available) is something different to experimental LCOW in moby from this Epic?

**Comment by jhowardmsft** [comment]
If you're referring to Docker for Windows, currently it can use this LCOW rather than moby VM as an experimental feature in D4W, but I understand "they" (folks at Docker Inc) are working on an alternate solution using WSL2.

**Comment by ggirard07** [comment]
@jhowardmsft regarding Docker announcement for WSL2, does it mean LCOW is going to die and be replaced by WSL2 support if conclusive?

**Comment by oldium** [comment]
Yes, I am referring to https://engineering.docker.com/2019/06/docker-hearts-wsl-2/, which is announced by Docker Inc folks. Anyway, using any solution will be great improvement over Samba connection to local files (in sense of reliability).

**Comment by krshubham** [comment]
Hi, is there a time line for a stable GA of LCOW?

**Comment by thaJeztah** [comment]
LCOW support is still experimental, and the current implementation is not recommended for serious work as there's still many features missing, and quite some known bugs/limitations; the current [will be removed at some point](https://github.com/moby/moby/issues/39533) (Active work is ongoing to re-implement LCOW at the [runtime level through containerd](https://github.com/moby/moby/pull/38541)).

If your use-case is to run Linux containers on a Windows machine, you might be interested in the tech preview of Docker Desktop for WSL2;

- [docker engineering blog: Docker ❤️ WSL 2 – The Future of Docker Desktop for Windows](https://engineering.docker.com/2019/06/docker-hearts-wsl-2/)
- [docker blog: Get Ready for the Tech Preview of Docker Desktop for WSL 2](https://blog.docker.com/2019/07/docker-wsl2-tech-preview/)
- [docker blog: 5 Things to Try with Docker Desktop WSL 2 Tech Preview](https://blog.docker.com/2019/07/5-things-docker-desktop-wsl2-tech-preview/)

**Comment by kojoru** [comment]
@thaJeztah do you know if there is any alternative (in beta, planned or otherwise) to manage Windows and Linux containers simultaneously without engine switching? WSL2 builds do not allow this at the moment and behave similarly to "Use Linux containers" mode.

That's the main use case for us at the moment, and it's unclear how if at all this will be supported in the future.

**Comment by AmitHole** [comment]
@thaJeztah do you know how to run Linux container on Windows (LCOW), with --network=transparent.
Because I have linux container which I want to directly run on the host using the transparent driver. So I am running the linux container in the LCOW mode with transparent driver as network. With the below mentioned command.

docker run -d --network=Mytransparent --platform=linux --name dynamicui dynamicui:dev

But ran into the below mentioned issue, after running the above command - 

 docker run -d --network TeamedNet --platform linux --name dynamicui dynamicui:dev
47cf6329ea2fe7487245547f3f816a3a12b7d3901e4b65358814ae0f3cd36e1a
docker: Error response from daemon: container 47cf6329ea2fe7487245547f3f816a3a12b7d3901e4b65358814ae0f3cd36e1a encountered an error during CreateProcess: failure in a Windows system call: Unspecified error (0x80004005)
[Event Detail: failed to link resolv.conf file for adapter de3394a3-4b6d-4cb9-9e19-04307b13390b: link /etc/resolv.conf /tmp/base/etc/resolv.conf: invalid cross-device link Stack Trace:
github.com/Microsoft/opengcs/service/gcs/oslayer/realos.(*realOS).Link
        /go/src/github.com/Microsoft/opengcs/service/gcs/oslayer/realos/realos.go:232
github.com/Microsoft/opengcs/service/gcs/core/gcs.(*gcsCore).configureAdapterInNamespace
        /go/src/github.com/Microsoft/opengcs/service/gcs/core/gcs/networking.go:57
github.com/Microsoft/opengcs/service/gcs/core/gcs.(*gcsCore).ExecProcess
        /go/src/github.com/Microsoft/opengcs/service/gcs/core/gcs/gcs.go:369
github.com/Microsoft/opengcs/service/gcs/bridge.(*Bridge).execProcess
        /go/src/github.com/Microsoft/opengcs/service/gcs/bridge/bridge.go:582
github.com/Microsoft/opengcs/service/gcs/bridge.(*Bridge).(github.com/Microsoft/opengcs/service/gcs/bridge.execProcess)-fm
        /go/src/github.com/Microsoft/opengcs/service/gcs/bridge/bridge.go:236
github.com/Microsoft/opengcs/service/gcs/bridge.HandlerFunc.ServeMsg
        /go/src/github.com/Microsoft/opengcs/service/gcs/bridge/bridge.go:68
github.com/Microsoft/opengcs/service/gcs/bridge.(*Mux).ServeMsg
        /go/src/github.com/Microsoft/opengcs/service/gcs/bridge/bridge.go:139
github.com/Microsoft/opengcs/service/gcs/bridge.(*Bridge).ListenAndServe.func2.1
        /go/src/github.com/Microsoft/opengcs/service/gcs/bridge/bridge.go:315
runtime.goexit
        /usr/lib/go/src/runtime/asm_amd64.s:2361 Provider: 00000000-0000-0000-0000-000000000000] extra info: {"CommandArgs":["/bin/sh","-c","./Cognex.DynamicUIManager.Server"],"WorkingDirectory":"/DynamicUIManager","Environment":{"ASPNETCORE_URLS":"http://+:80","ASPNETCORE_VERSION":"2.2.6","DOTNET_RUNNING_IN_CONTAINER":"true","HOSTNAME":"47cf6329ea2f","PATH":"/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"},"CreateStdInPipe":true,"CreateStdOutPipe":true,"CreateStdErrPipe":true,"ConsoleSize":[0,0],"OCISpecification":{"ociVersion":"1.0.1-dev","process":{"user":{"uid":0,"gid":0},"args":["/bin/sh","-c","./Cognex.DynamicUIManager.Server"],"env":["PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin","HOSTNAME=47cf6329ea2f","ASPNETCORE_URLS=http://+:80","DOTNET_RUNNING_IN_CONTAINER=true","ASPNETCORE_VERSION=2.2.6"],"cwd":"/DynamicUIManager","capabilities":{"bounding":["CAP_CHOWN","CAP_DAC_OVERRIDE","CAP_FSETID","CAP_FOWNER","CAP_MKNOD","CAP_NET_RAW","CAP_SETGID","CAP_SETUID","CAP_SETFCAP","CAP_SETPCAP","CAP_NET_BIND_SERVICE","CAP_SYS_CHROOT","CAP_KILL","CAP_AUDIT_WRITE"],"effective":["CAP_CHOWN","CAP_DAC_OVERRIDE","CAP_FSETID","CAP_FOWNER","CAP_MKNOD","CAP_NET_RAW","CAP_SETGID","CAP_SETUID","CAP_SETFCAP","CAP_SETPCAP","CAP_NET_BIND_SERVICE","CAP_SYS_CHROOT","CAP_KILL","CAP_AUDIT_WRITE"],"inheritable":["CAP_CHOWN","CAP_DAC_OVERRIDE","CAP_FSETID","CAP_FOWNER","CAP_MKNOD","CAP_NET_RAW","CAP_SETGID","CAP_SETUID","CAP_SETFCAP","CAP_SETPCAP","CAP_NET_BIND_SERVICE","CAP_SYS_CHROOT","CAP_KILL","CAP_AUDIT_WRITE"],"permitted":["CAP_CHOWN","CAP_DAC_OVERRIDE","CAP_FSETID","CAP_FOWNER","CAP_MKNOD","CAP_NET_RAW","CAP_SETGID","CAP_SETUID","CAP_SETFCAP","CAP_SETPCAP","CAP_NET_BIND_SERVICE","CAP_SYS_CHROOT","CAP_KILL","CAP_AUDIT_WRITE"]}},"root":{"path":"rootfs"},"hostname":"47cf6329ea2f","mounts":[{"destination":"/proc","type":"proc","source":"proc","options":["nosuid","noexec","nodev"]},{"destination":"/dev","type":"tmpfs","source":"tmpfs","options":["nosuid","strictatime","mode=755","size=65536k"]},{"destination":"/dev/pts","type":"devpts","source":"devpts","options":["nosuid","noexec","newinstance","ptmxmode=0666","mode=0620","gid=5"]},{"destination":"/sys","type":"sysfs","source":"sysfs","options":["nosuid","noexec","nodev","ro"]},{"destination":"/sys/fs/cgroup","type":"cgroup","source":"cgroup","options":["ro","nosuid","noexec","nodev"]},{"destination":"/dev/mqueue","type":"mqueue","source":"mqueue","options":["nosuid","noexec","nodev"]},{"destination":"/dev/shm","type":"tmpfs","source":"shm","options":["nosuid","noexec","nodev","mode=1777"]}],"linux":{"resources":{},"namespaces":[{"type":"mount"},{"type":"network"},{"type":"uts"},{"type":"pid"},{"type":"ipc"}],"maskedPaths":["/proc/asound","/proc/acpi","/proc/kcore","/proc/keys","/proc/latency_stats","/proc/timer_list","/proc/timer_stats","/proc/sched_debug","/proc/scsi","/sys/firmware"],"readonlyPaths":["/proc/bus","/proc/fs","/proc/irq","/proc/sys","/proc/sysrq-trigger"]},"windows":{"layerFolders":["C:\\ProgramData\\Docker\\lcow\\ce5656471f1972d1fb065863f34ca3d631c431f3fad97d77ff6fa6fd4bfb5f9b","C:\\ProgramData\\Docker\\lcow\\73c2b07e7168f0701ed45af4933f92ce9888a0b556014fd267d2d8f473618ed2","C:\\ProgramData\\Docker\\lcow\\7cb5da704e42d7ad33560322d8a69a2f59811501f288e5c7c238be792678a205","C:\\ProgramData\\Docker\\lcow\\284964a9ea03c7b4c044df2c54737000596d144a32c200a5dcf1d6a520abfbc5","C:\\ProgramData\\Docker\\lcow\\2657b3efd2cebb278566ff8a523aa8c9d7b83a841de626f461b5513a55b66092","C:\\ProgramData\\Docker\\lcow\\b1b031fdab0ab0279f303a8540a7eb04cf27bae2a08b1a8d82640b2c3601e79c","C:\\ProgramData\\Docker\\lcow\\5b8011f4e4693c51cd42000cc6afc6e127173f63a1746ee6c942487fe634d605","C:\\ProgramData\\Docker\\lcow\\2500e1d0c12cc21339dd2ae544337a08c38581afe160e0f7dd1bd4693f7378c3","C:\\ProgramData\\Docker\\lcow\\47cf6329ea2fe7487245547f3f816a3a12b7d3901e4b65358814ae0f3cd36e1a"],"hyperv":{},"network":{"endpointList":["de3394a3-4b6d-4cb9-9e19-04307b13390b"]}}}}.

**Comment by S93rastogi** [comment]
I am quite confused with the strategy of providing the Linux container support on Windows. I have Heard alot about LCOW and WSL to run linux container but I am still not sure why they are not providing the support of Linux Container using MOBY VM as in WIN 10 till things are not stable ( ie till LCOW is no more an experimental thing). There seems to be no fix strategy for the support of Linux containers on server 2016 and 2019. In server 2019 I have no other way except running linux container as LCOW and that is experimental. On server 2016 there is another issue if I run docker for win on it I don't get the support of the latest Docker builds as the new versions of Docker are not supported.
As there no solid way to run LINUX container on server 2016 and server 2019 that always works with no issues.

**Comment by Kralizek** [comment]
> I am quite confused with the strategy of providing the Linux container support on Windows. I have Heard alot about LCOW and WSL to run linux container but I am still not sure why they are not providing the support of Linux Container using MOBY VM as in WIN 10 till things are not stable ( ie till LCOW is no more an experimental thing). There seems to be no fix strategy for the support of Linux containers on server 2016 and 2019. In server 2019 I have no other way except running linux container as LCOW and that is experimental. On server 2016 there is another issue if I run docker for win on it I don't get the support of the latest Docker builds as the new versions of Docker are not supported.
> As there no solid way to run LINUX container on server 2016 and server 2019 that always works with no issues.

I would dare to say that the general perception is to use linux servers for linux containers (in production). LCOW and Moby are more developer tools than production tools.

**Comment by S93rastogi** [comment]
> > I am quite confused with the strategy of providing the Linux container support on Windows. I have Heard alot about LCOW and WSL to run linux container but I am still not sure why they are not providing the support of Linux Container using MOBY VM as in WIN 10 till things are not stable ( ie till LCOW is no more an experimental thing). There seems to be no fix strategy for the support of Linux containers on server 2016 and 2019. In server 2019 I have no other way except running linux container as LCOW and that is experimental. On server 2016 there is another issue if I run docker for win on it I don't get the support of the latest Docker builds as the new versions of Docker are not supported.
> > As there no solid way to run LINUX container on server 2016 and server 2019 that always works with no issues.
> 
> I would dare to say that the general perception is to use linux servers for linux containers (in production). LCOW and Moby are more developer tools than production tools.

One could not directly say that running Linux containers on Windows Environment is a wrong Strategy. Windows itself is putting efforts in this direction but the problem is that things are not clear. We also have client constraints that puts us to Windows Route

**Comment by BretFisher** [comment]
I agree that for servers, I see nothing from Microsoft or Docker to
indicate that “Linux on Windows” is a strategy. If you see documentation
indicating otherwise, please let me know. The strategy for both sides for
years now has been to make it easier for both windows and Linux containers
on windows 10 only.

**Comment by PatrickLang** [comment]
just a quick note - I haven't been tracking status of this since I changed teams early last year and am focusing on Kubernetes work. I'm looking for someone to give an update on what's next.

Personally I'm quite happy using WSL2 for development since I can build, test and run Linux code there including Moby, ContainerD and Kubernetes.

**Comment by oldium** [comment]
There was no update to vendor.conf (opengcs, hcsshim) for 7 months now. Is there anything to show to public, please?

**Comment by PatrickLang** [comment]
Linux Containers on Windows development moved to [ContainerD](https://github.com/containerd/containerd), and no more work is happening in the Moby codebase. I don't think we need to keep this epic open any longer since there's no progress tracked here.

cc @mamezgeb

**Comment by conioh** [comment]
@PatrickLang, cool. Where's the epic issue on the containerd repo tracking progress? I couldn't find it.
